src/Microsoft.Extensions.FileProviders.Physical/Watcher/DotnetFileWatcher.cs

```diff
@@ -0,0 +1,133 @@
+
+using System;
+using System.IO;
+
+namespace Microsoft.Extensions.FileProviders.Physical.Watcher
+{
+    internal class DotnetFileWatcher : IFileSystemWatcher
+    {
+        private readonly Func<string, FileSystemWatcher> _watcherFactory;
+        private readonly string _watchedDirectory;
+
+        private FileSystemWatcher _fileSystemWatcher;
+
+        public DotnetFileWatcher(string watchedDirectory)
+            : this(watchedDirectory, DefaultWatcherFactory)
+        {
+        }
+
+        internal DotnetFileWatcher(string watchedDirectory, Func<string, FileSystemWatcher> fileSystemWatcherFactory)
+        {
+            if (string.IsNullOrEmpty(watchedDirectory))
+            {
+                throw new ArgumentNullException(nameof(watchedDirectory));
+            }
+
+            _watchedDirectory = watchedDirectory;
+            _watcherFactory = fileSystemWatcherFactory;
+            CreateFileSystemWatcher();
+        }
+
+        public event EventHandler<string> OnFileChange;
+
+        public event EventHandler OnError;
+
+        private static FileSystemWatcher DefaultWatcherFactory(string watchedDirectory)
+        {
+            if (string.IsNullOrEmpty(watchedDirectory))
+            {
+                throw new ArgumentNullException(nameof(watchedDirectory));
+            }
+
+            return new FileSystemWatcher(watchedDirectory);
+        }
+
+        private void FSW_Error(object sender, ErrorEventArgs e)
+        {
+            // Recreate the watcher
+            CreateFileSystemWatcher();
+
+            if (OnError != null)
+            {
+                OnError(this, null);
+            }
+        }
+
+        private void FSW_Renamed(object sender, RenamedEventArgs e)
+        {
+            NotifyChange(e.OldFullPath);
+            NotifyChange(e.FullPath);
+
+            if (Directory.Exists(e.FullPath))
+            {
+                // If the renamed entity is a directory then notify tokens for every sub item.
+                foreach (var newLocation in Directory.EnumerateFileSystemEntries(e.FullPath, "*", SearchOption.AllDirectories))
+                {
+                    // Calculated previous path of this moved item.
+                    var oldLocation = Path.Combine(e.OldFullPath, newLocation.Substring(e.FullPath.Length + 1));
+                    NotifyChange(oldLocation);
+                    NotifyChange(newLocation);
+                }
+            }
+        }
+
+        private void FSW_Changed(object sender, FileSystemEventArgs e)
+        {
+
+            NotifyChange(e.FullPath);
+        }
+
+        private void NotifyChange(string fullPath)
+        {
+            if (OnFileChange != null)
+            {
+                // Only report file changes
+                OnFileChange(this, fullPath);
+            }
+        }
+
+        private void CreateFileSystemWatcher()
+        {
+            bool enableEvents = false;
+
+            if (_fileSystemWatcher != null)
```

Comment:
Are you guaranteed that this will never be called from multiple threads at the same time? (i.e. can it happen that there are two errors chasing each other?)